### DISCUSSION TITLE
selectAtom write functionality

### DISCUSSION THREAD

**Discussion by geril07** [pmndrs/jotai#2209] [Ideas] [answered]
I think would be cool to have selectAtom which can have writer(setter) for value which you select. 
For example:
```
const someAtom = atom({ a : 1, b : 2})
const bAtom = selectAtom(someAtom, (some) => some.b)
```
And you can have writer `bAtom`:
```
const store = getDefaultStore()/useStore() // just to call
store.set(bAtom, 10)
```
or 
```
const setBAtom = useSetAtom(bAtom)
setBatom(10)
```
And in this case `someAtom` would be `{a :1, b : 10}`.

Or maybe provide another argument to `selectAtom` where you can pass write function.

**Comment by dai-shi** [comment, marked as answer]
Here's some notes:

### Do not use `selectAtom` unless really necessary

Unlike its name, `selectAtom` is provided as an escape hatch. Using it means building not 100% pure atom tree.
Prefer using derived atoms and use selectAtom only when equalityFn or prevSlice is unavoidable.

If you use a normal derived atom, you can pass write function.

```js
const someAtom = atom({ a : 1, b : 2})
const bAtom = atom(
  (get) => get(someAtom).b,
  (get, set, newValue) => set(someAtom, (prev) => ({ ...prev, b: newValue }))
)
```

### Prefer composition even if with selectAtom

Jotai provides minimal utils and let developers create atoms in userland.

```js
const someAtom = atom({ a : 1, b : 2})
const internalAtom = selectAtom(someAtom, (some) => some.b)
const bAtom = atom(
  (get) => get(internalAtom),
  (get, set, newValue) => set(someAtom, (prev) => ({ ...prev, b: newValue }))
)
```

### btw, focusAtom might be of interest

`focusAtom` is provided by `jotai-optics` package.

```js
const someAtom = atom({ a : 1, b : 2})
const bAtom = focusAtom(someAtom, (optic) => optic.prop('b'))
```

**Reply by AjaxSolutions** [reply]
> Prefer using derived atoms and use selectAtom only when equalityFn or prevSlice is unavoidable.

This should be documented.

**Reply by dai-shi** [reply]
#2210

**Comment by geril07** [comment]
Thanks for the clarification.
